File: tp_tda/entregables/8/main.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
using namespace std;

map<int, vector<int>> adjacencies(vector<pair<int, int>> &edgeList) {
    map<int, vector<int>> adjacencyList;
    for (auto &edge : edgeList) {
        int node1 = edge.first, node2 = edge.second;
        adjacencyList[node1].push_back(node2);
        adjacencyList[node2].push_back(node1);
    }
    return adjacencyList;
}

void DFS(map<int, vector<int>> &adjacencyList, int start, map<int, bool> &visited) {
    vector<int> stack = {start};
    visited[start] = true;
    while (!stack.empty()) {
        int actualNode = stack.back();
        stack.pop_back();
        for (auto neighbor = adjacencyList[actualNode].rbegin(); neighbor != adjacencyList[actualNode].rend(); ++neighbor) {
            if (!visited[*neighbor]) {
                stack.push_back(*neighbor);
                visited[*neighbor] = true;
            }
        }
    }
}

int countComponents(map<int, vector<int>> &adjacencyList, map<int, bool> &visited) {
    int count = 0;
    for (auto &vertex : adjacencyList) {
        if (!visited[vertex.first]) {
            DFS(adjacencyList, vertex.first, visited);
            count++;
        }
    }
    return count;
}

map<int, vector<int>> deleteVertex(map<int, vector<int>> adjacencyList, int vertex) {
    adjacencyList.erase(vertex);
    for (auto &node : adjacencyList) {
        node.second.erase(remove(node.second.begin(), node.second.end(), vertex), node.second.end());
    }
    return adjacencyList;
}

void dfs2(map<int, vector<int>> &adjacencyList, int start, map<int, int> &parents, map<int, bool> &visited,
          map<int, int> &orden, map<int, int> &low, set<int> &artPoints, int &time) {
    visited[start] = true;
    orden[start] = low[start] = time++;
    int children = 0;

    for (int neighbor : adjacencyList[start]) {
        if (!visited[neighbor]) {
            parents[neighbor] = start;
            children++;
            dfs2(adjacencyList, neighbor, parents, visited, orden, low, artPoints, time);
            low[start] = min(low[start], low[neighbor]);

            if (parents[start] == -1 && children > 1)
                artPoints.insert(start);

            if (parents[start] != -1 && low[neighbor] >= orden[start])
                artPoints.insert(start);
        } else if (neighbor != parents[start]) {
            low[start] = min(low[start], orden[neighbor]);
        }
    }
}

vector<pair<int, int>> cutVertices(map<int, vector<int>> &adjacencyList, int k) {
    int n = adjacencyList.size();
    map<int, bool> visited;
    map<int, int> parents, orden, low;
    set<int> artPoints;
    int time = 0;

    for (auto &vertex : adjacencyList) {
        if (!visited[vertex.first]) {
            dfs2(adjacencyList, vertex.first, parents, visited, orden, low, artPoints, time);
        }
    }

    vector<pair<int, int>> output;
    for (int vertex : artPoints) {
        auto graph = adjacencyList;
        visited.clear();
        auto modified_graph = deleteVertex(graph, vertex);
        int components = countComponents(modified_graph, visited);
        output.push_back({vertex, components});
    }

    for (auto &vertex : adjacencyList) {
        if (output.size() == k)
            break;
        else if (artPoints.find(vertex.first) == artPoints.end()) {
            output.push_back({vertex.first, 1});
        }
    }

    sort(output.begin(), output.end(), [](const pair<int, int> &a, const pair<int, int> &b) {
        return (a.second > b.second) || (a.second == b.second && a.first < b.first);
    });

    return output;
}

int main() {
    
    ios_base::sync_with_stdio(0);
	cin.tie(0);
    
    vector<pair<vector<pair<int, int>>, int>> railwayEmpires;
    int vertex, cases;
    cin >> vertex >> cases;

    while (vertex != 0 && cases != 0) {
        vector<pair<int, int>> railway;
        int u, v;
        while (true) {
            cin >> u >> v;
            if (u == -1 && v == -1) break;
            railway.push_back({u, v});
        }
        railwayEmpires.push_back({railway, cases});
        cin >> vertex >> cases;
    }

    for (auto &railway : railwayEmpires) {
        auto adjacencyList = adjacencies(railway.first);
        auto cutVertex = cutVertices(adjacencyList, railway.second);
        for (int i = 0; i < railway.second; ++i) {
            cout << cutVertex[i].first << " " << cutVertex[i].second << endl;
        }
        cout << endl;
    }

    return 0;
}

File: tp_tda/entregables/8/main.py
def adjacencies(edgeList): # Función que recibe una lista de aristas y devuvelve una lista de adyacencias
    adjacencyList = {}
    for nodo1, nodo2 in edgeList:
        if nodo1 not in adjacencyList:
            adjacencyList[nodo1] = []
        if nodo2 not in adjacencyList:
            adjacencyList[nodo2] = []
        adjacencyList[nodo1].append(nodo2)
        adjacencyList[nodo2].append(nodo1)
    return adjacencyList

def DFS(adjacencyList, start, visited): # Función DFS
    stack = [start]
    visited[start] = True
    while stack != []:
        actualNode = stack.pop()
        for neighbor in reversed(adjacencyList[actualNode]):
            if not visited[neighbor]:
                stack.append(neighbor)
                visited[neighbor] = True

def countComponents(adjacencyList, visited): # Función que me cuentas las componentes del grafo
    count = 0
    for vertex in adjacencyList.keys():
        if not visited[vertex]:
            DFS(adjacencyList, vertex, visited)
            count += 1
    return count

def deleteVertex(adjacencyList, vertex): # Función que dado una lista de adyacencias y un vertice, me elimina ese vertice de todo el grafo
    del adjacencyList[vertex]
    for node in adjacencyList:
        adjacencyList[node] = list(set(adjacencyList[node]) - {vertex})
    return adjacencyList

def dfs2(adjacencyList, start, parents, visited, orden, low, artPoints, time): # Variante de DFS que mira los puntos de articulación
    visited[start] = True
    orden[start] = low[start] = time[0]
    time[0] += 1
    children = 0

    for neighbor in adjacencyList[start]:
        if not visited[neighbor]:
            parents[neighbor] = start
            children += 1
            dfs2(adjacencyList, neighbor, parents, visited, orden, low, artPoints, time)
            low[start] = min(low[start], low[neighbor])

            if parents[start] == -1 and children > 1:
                artPoints.add(start)

            if parents[start] != -1 and low[neighbor] >= orden[start]:
                artPoints.add(start)
        elif neighbor != parents[start]:
            low[start] = min(low[start], orden[neighbor])

def cutVertices(adjacencyList, k):
    n = len(adjacencyList)
    visited = [False] * n
    parents = [-1] * n
    orden = [0] * n
    low = [0] * n
    artPoints = set()
    time = [0]

    for vertex in range(n):
        if not visited[vertex]:
            dfs2(adjacencyList, vertex, parents, visited, orden, low, artPoints, time) # Busco todos los puntos de articulacion
    output = []
    for vertex in artPoints: # Por cada punto de articulacion lo elimino del grafo y cuento las componentes que hay
        graph = adjacencyList.copy()
        visited = [False]*n
        modified_graph = deleteVertex(graph, vertex)
        output.append((vertex, countComponents(modified_graph, visited)))
    for vertex in range(n):
        if (len(output) == k):
            break
        else:
            if (vertex not in artPoints):
                output.append((vertex, 1))
    output = sorted(output, key=lambda x: (-x[1], x[0]))
    return output

# Tome idea del codigo que hicieron en el turno noche

def main():
    # Guardo los inputs
    railwayEmpires = []
    data = list(map(int, input().split()))
    cases = data[1]
    vertex = data[0]
    while (data[0] != 0 and data[1] != 0):
        railway = []
        while (data[0] != -1 and data[1] != -1):
            data = list(map(int, input().split()))
            if (data[0] != -1 and data[1] != -1):
                edge = (data[0], data[1])
                railway.append(edge)
        railwayEmpires.append((railway, cases))
        data = list(map(int, input().split()))
        cases = data[1]
        vertex = data[0]
    # Termino de guardar el input
    for railway in railwayEmpires:
        adjacencyList = adjacencies(railway[0]) # Paso a una lista de adyacencia
        cutVertex = cutVertices(adjacencyList, railway[1]) # Calculo los puntos de articulacion y cuantas componentes generan si los elimino
        for i in range(railway[1]): # Imprimo los resultados
            print(f"{cutVertex[i][0]} {cutVertex[i][1]}")
        print("")

main()